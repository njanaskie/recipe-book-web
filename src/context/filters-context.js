import React, { useContext, useReducer } from 'react';
import filtersReducer, { filtersReducerDefaultState } from '../reducers/filters'

const FiltersContext = React.createContext()

export const useFiltersContext = () => useContext(FiltersContext)

const FiltersProvider = ({ children }) => {
    const [filters, filtersDispatch] = useReducer(filtersReducer, filtersReducerDefaultState)

    const filterIngredients = (ingredients) => {
        filtersDispatch({ type: 'SET_INGREDIENT_FILTER', ingredients })
    }

    const filterCuisine = (cuisine) => {
        filtersDispatch({ type: 'SET_CUISINE_FILTER', cuisine })
    }

    const filterType = (recipeType) => {
        filtersDispatch({ type: 'SET_RECIPE_TYPE_FILTER', recipeType })
    }

    const filterCustomTags = (customTags) => {
        filtersDispatch({ type: 'SET_CUSTOM_TAG_FILTER', customTags })
    }

    return (
        <FiltersContext.Provider value={{ filters, filtersDispatch, filterCuisine, filterIngredients, filterType, filterCustomTags }}>
            {children}
        </FiltersContext.Provider>
    )
}

export { FiltersProvider, FiltersContext as default }