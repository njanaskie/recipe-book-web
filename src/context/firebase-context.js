import React, { useEffect, useState, useContext, useReducer } from 'react';
import database, { firebase, googleAuthProvider } from '../firebase/firebase';
import { history } from '../routers/AppRouter';
import authReducer from '../reducers/auth'
import LoadingPage from '../components/LoadingPage'

const FirebaseContext = React.createContext()

export const useFirebaseContext = () => useContext(FirebaseContext)

const FirebaseProvider = ({ children }) => {
    const [auth, authDispatch] = useReducer(authReducer, [])
    const [user, setUser] = useState('')
    const [loading, setLoading] = useState(true)
    const [isAdmin, setIsAdmin] = useState()
    const [isGuest, setIsGuest] = useState()
    const email = process.env.GUEST_EMAIL
    const password = process.env.GUEST_PASSWORD

    const login = () => {
        firebase.auth().signInWithPopup(googleAuthProvider).then(() => {
            authDispatch({ type: 'LOGIN' })
        })
    }

    const loginAsGuest = () => {
        firebase.auth().signInWithEmailAndPassword(email, password).then(() => {
            authDispatch({ type: 'LOGIN_AS_GUEST' })
        })
    }

    const logout = () => {
        firebase.auth().signOut().then(() => {
            authDispatch({ type: 'LOGOUT' })
        });
    }

    useEffect(() => {
        const unsubscribe = firebase.auth().onAuthStateChanged((user) => {
            setUser(user)
            setLoading(false)
            setIsAdmin()
            setIsGuest()
        })

        return () => unsubscribe()
    }, [])

    useEffect(() => {
        if (user) {
            database.collection('users').doc(user.uid).get()
            .then((doc) => {
                const userData = doc.data()
                if (userData) {
                    if (userData.isAdmin === true) {
                        const isAdmin = userData.isAdmin
                        setIsAdmin(isAdmin)
                    } else if (userData.isGuest === true) {
                        const isGuest = userData.isGuest
                        setIsGuest(isGuest)
                    }
                }
            })
        }
    }, [user])

    if (loading) {
        return <LoadingPage />
    }
    // if (user) {
    //     if (history.location.pathname === '/') {
    //         history.push('/home')
    //     }
    // } else {
    //     history.push('/');
    // }

    return (
        <FirebaseContext.Provider value={{ user, loading, isAdmin, isGuest, login, loginAsGuest, logout }}>
            {children}
        </FirebaseContext.Provider>
    )
}

export { FirebaseProvider, FirebaseContext as default }