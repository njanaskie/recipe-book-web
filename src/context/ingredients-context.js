import React, { useContext, useReducer } from 'react';
import ingredientsReducer from '../reducers/ingredients'
import database from '../firebase/firebase'

const IngredientsContext = React.createContext()

export const useIngredientsContext = () => useContext(IngredientsContext)

const IngredientsProvider = ({ children }) => {
    const [ingredients, dispatch ] = useReducer(ingredientsReducer, [])

    const removeIngredient = ({ id } = {}) => {
        return database.collection('ingredients').doc(id).delete().then(() => {
            dispatch({ type: 'REMOVE_INGREDIENT', id })
        });
    };

    const addIngredient = (ingredient) => {
        database.collection('ingredients').add(ingredient).then((ref) => {
            dispatch(({ type: 'ADD_INGREDIENT', ingredient: {id: ref.key, ...ingredient} }))
        })
    }

    return (
        <IngredientsContext.Provider value={{ ingredients, dispatch, removeIngredient, addIngredient }}>
            {children}
        </IngredientsContext.Provider>
    )
}

export { IngredientsProvider, IngredientsContext as default }