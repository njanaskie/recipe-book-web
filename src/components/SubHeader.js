import React, { useState } from 'react';
import { NavLink } from 'react-router-dom';
import { Button, Menu, Icon } from 'semantic-ui-react'
import { useFirebaseContext } from '../context/firebase-context'
import RecipeInputModal from './RecipeInputModal';
import AddRecipe from './AddRecipe'

export const SubHeader = () => {
    const { isAdmin, logout } = useFirebaseContext()
    const [isModalOpen, setIsModalOpen] = useState(false)

    const handleModalOpen = () => {
        setIsModalOpen(true)
    }

    const handleModalClose = () => {
        setIsModalOpen(false)
    }

    const handleLogout = () => {
        logout()
    }

    return (
        <div>
            <Menu secondary stackable>
                {isAdmin && <Menu.Item as={NavLink} to='/ingredients'><Icon name='lemon'/>Manage Ingredients</Menu.Item>}
                <Menu.Item as={Button} onClick={handleModalOpen}><Icon name='plus'/>Add Recipe</Menu.Item>
                <Menu.Item><Button onClick={handleLogout}>Logout</Button></Menu.Item>
            </Menu>
            <RecipeInputModal isModalOpen={isModalOpen} handleModalClose={handleModalClose}>
                <AddRecipe handleModalClose={handleModalClose}/>
            </RecipeInputModal>
        </div>
    )
}

export default SubHeader;