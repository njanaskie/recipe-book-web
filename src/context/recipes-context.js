import React, { useContext, useReducer } from 'react';
import recipesReducer from '../reducers/recipes'
import database from '../firebase/firebase'
import uuid from 'uuid'
import { useFirebaseContext } from './firebase-context'


const RecipesContext = React.createContext()

export const useRecipesContext = () => useContext(RecipesContext)

const RecipesProvider = ({ children }) => {
    const { user } = useFirebaseContext() 
    const [recipes, recipeDispatch] = useReducer(recipesReducer, [])
    const customId = uuid()

    const addRecipe = (recipe) => {
        database.collection('users').doc(user.uid).collection('recipes').doc(customId).set(recipe).then(() => {
            recipeDispatch({ type: 'ADD_RECIPE', recipe: {id: customId, ...recipe} })
        })
    }

    const editRecipe = (id, updates) => {
        database.collection('users').doc(user.uid).collection('recipes').doc(id).update(updates).then(() => {
            recipeDispatch({ type: 'EDIT_RECIPE', id, updates })
        })
    }

    const removeRecipe = (id) => {
        database.collection('users').doc(user.uid).collection('recipes').doc(id).delete().then(() => {
            recipeDispatch({ type: 'REMOVE_RECIPE', id })
        })
    }

    return (
        <RecipesContext.Provider value={{ recipes, recipeDispatch, addRecipe, editRecipe, removeRecipe }}>
            {children}
        </RecipesContext.Provider>
    )
}

export { RecipesProvider, RecipesContext as default }