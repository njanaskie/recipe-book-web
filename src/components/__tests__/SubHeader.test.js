import React from 'react'
import { shallow } from 'enzyme'
import { SubHeader } from '../SubHeader'
import * as FirebaseContext from '../../context/firebase-context'

test('should render SubHeader when admin exists',() => {
    jest.spyOn(FirebaseContext, 'useFirebaseContext').mockImplementation(() => ({isAdmin: true}))
    const wrapper = shallow(<SubHeader />)
    expect(wrapper).toMatchSnapshot()
});

test('should render SubHeader when admin does not exist',() => {
    jest.spyOn(FirebaseContext, 'useFirebaseContext').mockImplementation(() => ({isAdmin: false}))
    const wrapper = shallow(<SubHeader />)
    expect(wrapper).toMatchSnapshot()
});