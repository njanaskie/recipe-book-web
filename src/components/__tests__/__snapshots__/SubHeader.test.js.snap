// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render SubHeader when admin does not exist 1`] = `
<header>
  <NavLink
    activeClassName="is-active"
    to="/dashboard"
  >
    Dashboard
  </NavLink>
  <NavLink
    activeClassName="is-active"
    to="/pantry"
  >
    Pantry
  </NavLink>
</header>
`;

exports[`should render SubHeader when admin exists 1`] = `
<header>
  <NavLink
    activeClassName="is-active"
    to="/dashboard"
  >
    Dashboard
  </NavLink>
  <NavLink
    activeClassName="is-active"
    to="/pantry"
  >
    Pantry
  </NavLink>
  <NavLink
    activeClassName="is-active"
    to="/ingredients"
  >
    Ingredients
  </NavLink>
  <NavLink
    activeClassName="is-active"
    to="/dishes"
  >
    Dishes
  </NavLink>
  <NavLink
    activeClassName="is-active"
    to="/add-dish"
  >
    Add Dish
  </NavLink>
</header>
`;
